spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -368,13 +349,25 @@ trait TableFeatureSupport { this: Protocol =>
    * features. After we remove the last native feature we downgrade the protocol to (1, 1).
    */
   def downgradeProtocolVersionsIfNeeded: Protocol = {
-    if (!readerAndWriterFeatures.forall(_.isLegacyFeature)) return this
+    if (nativeReaderAndWriterFeatures.nonEmpty) {
+      val (minReaderVersion, minWriterVersion) =
+        TableFeatureProtocolUtils.minimumRequiredVersions(nativeReaderAndWriterFeatures)
+      // It is guaranteed by the definitions of WriterFeature and ReaderFeature, that we cannot
+      // end up with invalid protocol versions such as (3, 3). Nevertheless,
+      // we double check it here.
+      val newProtocol =
+        Protocol(minReaderVersion, minWriterVersion).withFeatures(readerAndWriterFeatures)
+      assert(
+        newProtocol.supportsWriterFeatures,
+        s"Downgraded protocol should at least support writer features, but got $newProtocol.")
+      return newProtocol
+    }
```

Comment:
IIUIC `(minReaderVersion, minWriterVersion)` will end up being either (1, 7) or (3, 7) depending on whether the `nativeReaderAndWriterFeatures` contain any reader features. If the ColumnMapping feature is present, and a native writer feature, then this will return `(1, 7)`, but then the `newProtocol` below will use `withFeatures` to add back ColumnMapping. I think that would fail because `withFeatures` can't upgrade the legacy protocol version numbers.

Should this have been `readerAndWiterFeatures` in the third line? In that case we would have ended up with (2, 7). And then `withFeatures` works for adding ColumnMapping, because it accepts legacy features that are already implicitly supported.

Let's make sure we have a test for downgrading to (2, 7).

Reply:
Yes. I thought that once a table is in the table features territory there should be no interaction with legacy features when it comes to protocol versions. But that is incorrect. Fixed and added a test.